doc/man/raptoreumd.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.11.
.TH RAPTOREUMD "1" December 2021" "raptoreumd v0.17.0.0" "User Commands"
.SH NAME
raptoreumd \- manual page for raptoreumd v0.17.0.0
.SH DESCRIPTION
Raptoreum Core Daemon version v0.17.0.0
.SS "Usage:"
.TP
raptoreumd [options]
Start Raptoreum Core Daemon
.SH OPTIONS
.HP
\-?
.IP
Print this help message and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when a relevant alert is received or we see a really
long fork (%s in cmd is replaced by message)
.HP
\fB\-assumevalid=\fR<hex>
.IP
If this block is in the chain assume that it and its ancestors are valid
and potentially skip their script verification (0 to verify all,
default:
00000000000000105cfae44a995332d8ec256850ea33a1f7b700474e3dad82bc,
testnet:
0000009303aeadf8cf3812f5c869691dbd4cb118ad20e9bf553be434bafe6a52)
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-blockreconstructionextratxn=\fR<n>
.IP
Extra transactions to keep in memory for compact block reconstructions
(default: 100)
.HP
\fB\-blocksdir=\fR<dir>
.IP
Specify blocks directory (default: <datadir>/blocks)
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file. Relative paths will be prefixed by datadir
location. (default: raptoreum.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-dbcache=\fR<n>
.IP
Set database cache size in megabytes (4 to 16384, default: 300)
.HP
\fB\-debuglogfile=\fR<file>
.IP
Specify location of debug log file. Relative paths will be prefixed by a
net\-specific datadir location. (0 to disable; default: debug.log)
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external blk000??.dat file on startup
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the transaction memory pool below <n> megabytes (default: 300)
.HP
\fB\-maxorphantxsize=\fR<n>
.IP
Maximum total size of all orphan transactions in megabytes (default: 10)
.HP
\fB\-maxrecsigsage=\fR<n>
.IP
Number of seconds to keep LLMQ recovery sigs (default: 604800)
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
336)
.HP
\fB\-par=\fR<n>
.IP
Set the number of script verification threads (\fB\-4\fR to 16, 0 = auto, <0 =
leave that many cores free, default: 0)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file. Relative paths will be prefixed by a net\-specific
datadir location. (default: raptoireumd.pid)
.HP
\fB\-prune=\fR<n>
.IP
Reduce storage requirements by enabling pruning (deleting) of old
blocks. This allows the pruneblockchain RPC to be called to
delete specific blocks, and enables automatic pruning of old
blocks if a target size in MiB is provided. This mode is
incompatible with \fB\-txindex\fR, \fB\-rescan\fR and \fB\-disablegovernance\fR=\fI\,false\/\fR.
Warning: Reverting this setting requires re\-downloading the
entire blockchain. (default: 0 = disable pruning blocks, 1 =
allow manual pruning via RPC, >945 = automatically prune block
files to stay under the specified target size in MiB)
.HP
\fB\-syncmempool\fR
.IP
Sync mempool from other nodes on start (default: 1)
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-version\fR
.IP
Print version and exit
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open (see
the `addnode` RPC command help for more info). This option can be
specified multiple times to add multiple nodes.
.HP
\fB\-allowprivatenet\fR
.IP
Allow RFC1918 addresses to be relayed and connected to (default: 0)
.HP
\fB\-banscore=\fR<n>
.IP
Threshold for disconnecting misbehaving peers (default: 100)
.HP
\fB\-bantime=\fR<n>
.IP
Number of seconds to keep misbehaving peers from reconnecting (default:
86400)
.HP
\fB\-bind=\fR<addr>
.IP
Bind to given address and always listen on it. Use [host]:port notation
for IPv6
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node; \fB\-connect\fR=\fI\,0\/\fR disables automatic
connections (the rules for this peer are the same as for
\fB\-addnode\fR). This option can be specified multiple times to connect
to multiple nodes.
.HP
\fB\-discover\fR
.IP
Discover own IP addresses (default: 1 when listening and no \fB\-externalip\fR
or \fB\-proxy\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR used)
.HP
\fB\-enablebip61\fR
.IP
Send reject messages per BIP61 (default: 1)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 0)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or \fB\-connect\fR)
.HP
\fB\-listenonion\fR
.IP
Automatically create Tor hidden service (default: 1)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (temporary service connections
excluded) (default: 125)
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer, <n>*1000 bytes (default: 5000)
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer, <n>*1000 bytes (default: 1000)
.HP
\fB\-maxtimeadjustment\fR
.IP
Maximum allowed median peer time offset adjustment. Local perspective of
time may be influenced by peers forward or backward by this
amount. (default: 4200 seconds)
.HP
\fB\-maxuploadtarget=\fR<n>
.IP
Tries to keep outbound traffic under the given target (in MiB per 24h),
0 = no limit (default: 0)
.HP
\fB\-onion=\fR<ip:port>
.IP
Use separate SOCKS5 proxy to reach peers via Tor hidden services
(default: \fB\-proxy\fR)
.HP
\fB\-onlynet=\fR<net>
.IP
Make outgoing connections only through network <net> (ipv4, ipv6 or
onion). Incoming connections are not affected by this option.
This option can be specified multiple times to allow multiple
networks.
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
1)
.HP
\fB\-peertimeout=\fR<n>
.IP
Specify p2p connection timeout in seconds. This option determines the
amount of time a peer may be inactive before the connection to it
is dropped. (minimum: 1, default: 60)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port> (default: 9999 or testnet: 19999)
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. This enables Tor
stream isolation (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect. This
option can be specified multiple times to connect to multiple
nodes.
.HP
\fB\-socketevents=\fR<mode>
.IP
Socket events mode, which must be one of 'select', 'poll', 'epoll' or
\&'kqueue', depending on your system (default: Linux \- 'epoll',
FreeBSD/Apple \- 'kqueue', Windows \- 'select')
.HP
\fB\-socketevents=\fR<mode>
.IP
Socket events mode, which must be one of: 'select' (default: select)
.HP
\fB\-timeout=\fR<n>
.IP
Specify connection timeout in milliseconds (minimum: 1, default: 5000)
.HP
\fB\-torcontrol=\fR<ip>:<port>
.IP
Tor control port to use if onion listening enabled (default:
127.0.0.1:9051)
.HP
\fB\-torpassword=\fR<pass>
.IP
Tor control port password (default: empty)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 0)
.HP
\fB\-whitebind=\fR<addr>
.IP
Bind to given address and whitelist peers connecting to it. Use
[host]:port notation for IPv6
.HP
\fB\-whitelist=\fR<IP address or network>
.IP
Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
CIDR notated network (e.g. 1.2.3.0/24). Can be specified multiple
times. Whitelisted peers cannot be DoS banned and their
transactions are always relayed, even if they are already in the
mempool, useful e.g. for a gateway
.PP
Indexing options:
.HP
\fB\-addressindex\fR
.IP
Maintain a full address index, used to query for the balance, txids and
unspent outputs for addresses (default: 0)
.HP
\fB\-reindex\fR
.IP
Rebuild chain state and block index from the blk*.dat files on disk
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks
.HP
\fB\-spentindex\fR
.IP
Maintain a full spent index, used to query the spending txid and input
index for an outpoint (default: 0)
.HP
\fB\-timestampindex\fR
.IP
Maintain a timestamp index for block hashes, used to query blocks hashes
by a range of timestamps (default: 0)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 1)
.PP
Masternode options:
.HP
\fB\-llmq\-data\-recovery=\fR<n>
.IP
Enable automated quorum data recovery (default: 1)
.HP
\fB\-llmq\-qvvec\-sync=\fR<quorum_name>:<mode>
.IP
Defines from which LLMQ type the masternode should sync quorum
verification vectors. Can be used multiple times with different
LLMQ types. <mode>: 0 (sync always from all quorums of the type
defined by <quorum_name>), 1 (sync from all quorums of the type
defined by <quorum_name> if a member of any of the quorums)
.HP
\fB\-masternodeblsprivkey=\fR<hex>
.IP
Set the masternode BLS private key and enable the client to act as a
masternode
.HP
\fB\-platform\-user=\fR<user>
.IP
Set the username for the "platform user", a restricted user intended to
be used by Dash Platform, to the specified username.
.PP
Statsd options:
.HP
\fB\-statsenabled\fR
.IP
Publish internal stats to statsd (default: 0)
.HP
\fB\-statshost=\fR<ip>
.IP
Specify statsd host (default: 127.0.0.1)
.HP
\fB\-statshostname=\fR<ip>
.IP
Specify statsd host name (default: )
.HP
\fB\-statsns=\fR<ns>
.IP
Specify additional namespace prefix (default: )
.HP
\fB\-statsperiod=\fR<seconds>
.IP
Specify the number of seconds between periodic measurements (default:
60)
.HP
\fB\-statsport=\fR<port>
.IP
Specify statsd port (default: 8125)
.PP
Wallet options:
.HP
\fB\-createwalletbackups=\fR<n>
.IP
Number of automatic wallet backups (default: 10)
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls
.HP
\fB\-instantsendnotify=\fR<cmd>
.IP
Execute command when a wallet InstantSend transaction is successfully
locked (%s in cmd is replaced by TxID)
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 1000)
.HP
\fB\-rescan=\fR<mode>
.IP
Rescan the block chain for missing wallet transactions on startup (1 =
start from wallet creation time, 2 = start from genesis block)
.HP
\fB\-salvagewallet\fR
.IP
Attempt to recover private keys from a corrupt wallet on startup
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-upgradewallet\fR
.IP
Upgrade wallet to latest format on startup
.HP
\fB\-wallet=\fR<path>
.IP
Specify wallet database path. Can be specified multiple times to load
multiple wallets. Path is interpreted relative to <walletdir> if
it is not absolute, and will be created if it does not exist (as
a directory containing a wallet.dat file and log files). For
backwards compatibility this will also accept names of existing
data files in <walletdir>.)
.HP
\fB\-walletbackupsdir=\fR<dir>
.IP
Specify full path to directory for automatic wallet backups (must exist)
.HP
\fB\-walletbroadcast\fR
.IP
Make the wallet broadcast transactions (default: 1)
.HP
\fB\-walletdir=\fR<dir>
.IP
Specify directory to hold wallets (default: <datadir>/wallets if it
exists, otherwise <datadir>)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes (%s in cmd is replaced
by TxID)
.HP
\fB\-zapwallettxes=\fR<mode>
.IP
Delete all wallet transactions and only recover those parts of the
blockchain through \fB\-rescan\fR on startup (1 = keep tx meta data e.g.
account owner and payment request information, 2 = drop tx meta
data)
.HP
\fB\-discardfee=\fR<amt>
.IP
The fee rate (in RAPTOREUM/kB) that indicates your tolerance for discarding
change by adding it to the fee (default: 0.0001). Note: An output
is discarded if it is dust at this rate, but we will always
discard up to the dust relay fee and a discard fee above that is
limited by the fee estimate for the longest target
.HP
\fB\-fallbackfee=\fR<amt>
.IP
A fee rate (in DASH/kB) that will be used when fee estimation has
insufficient data (default: 0.00001)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fees (in RAPTOREUM/kB) smaller than this are considered zero fee for
transaction creation (default: 0.00001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee (in RAPTOREUM/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-txconfirmtarget=\fR<n>
.IP
If paytxfee is not set, include enough fee so transactions begin
confirmation on average within n blocks (default: 6)
.HP
\fB\-hdseed=\fR<hex>
.IP
User defined seed for HD wallet (should be in hex). Only has effect
during wallet creation/first start (default: randomly generated)
.HP
\fB\-mnemonic=\fR<text>
.IP
User defined mnemonic for HD wallet (bip39). Only has effect during
wallet creation/first start (default: randomly generated)
.HP
\fB\-mnemonicpassphrase=\fR<text>
.IP
User defined mnemonic passphrase for HD wallet (BIP39). Only has effect
during wallet creation/first start (default: empty string)
.HP
\fB\-usehd\fR
.IP
Use hierarchical deterministic key generation (HD) after BIP39/BIP44.
Only has effect during wallet creation/first start (default: 0)
.HP
\fB\-keepass\fR
.IP
Use KeePass 2 integration using KeePassHttp plugin (default: 0)
.HP
\fB\-keepassid=\fR<id>
.IP
KeePassHttp id for the established association
.HP
\fB\-keepasskey=\fR<key>
.IP
KeePassHttp key for AES encrypted communication with KeePass
.HP
\fB\-keepassname=\fR<name>
.IP
Name to construct url for KeePass entry that stores the wallet
passphrase
.HP
\fB\-keepassport=\fR<port>
.IP
Connect to KeePassHttp on port <port> (default: 19455)
.HP
\fB\-coinjoinamount=\fR<n>
.IP
Target CoinJoin balance (2\-21000000, default: 1000)
.HP
\fB\-coinjoinautostart\fR
.IP
Start CoinJoin automatically (0\-1, default: 0)
.HP
\fB\-coinjoindenomsgoal=\fR<n>
.IP
Try to create at least N inputs of each denominated amount (10\-100000,
default: 50)
.HP
\fB\-coinjoindenomshardcap=\fR<n>
.IP
Create up to N inputs of each denominated amount (10\-100000, default:
300)
.HP
\fB\-coinjoinmultisession\fR
.IP
Enable multiple CoinJoin mixing sessions per block, experimental (0\-1,
default: 0)
.HP
\fB\-coinjoinrounds=\fR<n>
.IP
Use N separate masternodes for each denominated input to mix funds
(2\-16, default: 4)
.HP
\fB\-coinjoinsessions=\fR<n>
.IP
Use N separate masternodes in parallel to mix funds (1\-10, default: 4)
.HP
\fB\-enablecoinjoin\fR
.IP
Enable use of CoinJoin for funds stored in this wallet (0\-1, default: 0)
.PP
PrivateSend options:
.HP
\fB\-enableprivatesend\fR
.IP
Enable use of PrivateSend for funds stored in this wallet (0\-1, default:
0)
.HP
\fB\-privatesendautostart\fR
.IP
Start PrivateSend automatically (0\-1, default: 0)
.HP
\fB\-privatesendmultisession\fR
.IP
Enable multiple PrivateSend mixing sessions per block, experimental
(0\-1, default: 0)
.HP
\fB\-privatesendsessions=\fR<n>
.IP
Use N separate masternodes in parallel to mix funds (1\-10, default: 4)
.HP
\fB\-privatesendrounds=\fR<n>
.IP
Use N separate masternodes for each denominated input to mix funds
(2\-16, default: 4)
.HP
\fB\-privatesendamount=\fR<n>
.IP
Target PrivateSend balance (2\-21000000, default: 1000)
.HP
\fB\-privatesenddenomsgoal=\fR<n>
.IP
Try to create at least N inputs of each denominated amount (10\-100000,
default: 50)
.HP
\fB\-privatesenddenomshardcap=\fR<n>
.IP
Create up to N inputs of each denominated amount (10\-100000, default:
300)
.PP
ZeroMQ notification options:
.HP
\fB\-zmqpubhashblock=\fR<address>
.IP
Enable publish hash block in <address>
.HP
\fB\-zmqpubhashgovernanceobject=\fR<address>
.IP
Enable publish hash of governance objects (like proposals) in <address>
.HP
\fB\-zmqpubhashgovernancevote=\fR<address>
.IP
Enable publish hash of governance votes in <address>
.HP
\fB\-zmqpubhashinstantsenddoublespend=\fR<address>
.IP
Enable publish transaction hashes of attempted InstantSend double spend
in <address>
.HP
\fB\-zmqpubhashrecoveredsig=\fR<address>
.IP
Enable publish message hash of recovered signatures (recovered by LLMQs)
in <address>
.HP
\fB\-zmqpubhashtx=\fR<address>
.IP
Enable publish hash transaction in <address>
.HP
\fB\-zmqpubhashtxlock=\fR<address>
.IP
Enable publish hash transaction (locked via InstantSend) in <address>
.HP
\fB\-zmqpubrawblock=\fR<address>
.IP
Enable publish raw block in <address>
.HP
\fB\-zmqpubrawinstantsenddoublespend=\fR<address>
.IP
Enable publish raw transactions of attempted InstantSend double spend in
<address>
.HP
\fB\-zmqpubrawrecoveredsig=\fR<address>
.IP
Enable publish raw recovered signatures (recovered by LLMQs) in
<address>
.HP
\fB\-zmqpubrawtx=\fR<address>
.IP
Enable publish raw transaction in <address>
.HP
\fB\-zmqpubrawtxlock=\fR<address>
.IP
Enable publish raw transaction (locked via InstantSend) in <address>
.PP
Debugging/Testing options:
.HP
\fB\-debug=\fR<category>
.IP
Output debugging information (default: 0, supplying <category> is
optional). If <category> is not supplied or if <category> = 1,
output all debugging information. <category> can be: net, tor,
mempool, http, bench, zmq, db, rpc, estimatefee, addrman,
selectcoins, reindex, cmpctblock, rand, prune, proxy, mempoolrej,
libevent, coindb, qt, leveldb, chainlocks, gobject, instantsend,
keepass, llmq, llmq\-dkg, llmq\-sigs, mnpayments, mnsync, coinjoin,
spork, netconn.
.HP
\fB\-debugexclude=\fR<category>
.IP
Exclude debugging information for a category. Can be used in conjunction
with \fB\-debug\fR=\fI\,1\/\fR to output debug logs for all categories except one
or more specified categories.
.HP
\fB\-disablegovernance\fR
.IP
Disable governance validation (0\-1, default: 0)
.HP
\fB\-help\-debug\fR
.IP
Show all debugging options (usage: \fB\-\-help\fR \fB\-help\-debug\fR)
.HP
\fB\-highsubsidyblocks=\fR<n>
.IP
The number of blocks with a higher than normal subsidy to mine at the
start of a devnet (default: 0)
.HP
\fB\-highsubsidyfactor=\fR<n>
.IP
The factor to multiply the normal block subsidy by while in the
highsubsidyblocks window of a devnet (default: 1)
.HP
\fB\-llmqchainlocks=\fR<quorum name>
.IP
Override the default LLMQ type used for ChainLocks on a devnet. Allows
using ChainLocks with smaller LLMQs. (default: llmq_50_60)
.HP
\fB\-llmqdevnetparams=\fR<size:threshold>
.IP
Override the default LLMQ size for the LLMQ_DEVNET quorum (default:
10:6)
.HP
\fB\-llmqinstantsend=\fR<quorum name>
.IP
Override the default LLMQ type used for InstantSend on a devnet. Allows
using InstantSend with smaller LLMQs. (default: llmq_50_60)
.HP
\fB\-llmqtestparams=\fR<size:threshold>
.IP
Override the default LLMQ size for the LLMQ_TEST quorum (default: 3:2)
.HP
\fB\-logips\fR
.IP
Include IP addresses in debug output (default: 0)
.HP
\fB\-logtimestamps\fR
.IP
Prepend debug output with timestamp (default: 1)
.HP
\fB\-maxtxfee=\fR<amt>
.IP
Maximum total fees (in RAPTOREUM) to use in a single wallet transaction or
raw transaction; setting this too low may abort large
transactions (default: 0.10)
.HP
\fB\-minimumdifficultyblocks=\fR<n>
.IP
The number of blocks that can be mined with the minimum difficulty at
the start of a devnet (default: 0)
.HP
\fB\-minsporkkeys=\fR<n>
.IP
Overrides minimum spork signers to change spork value. Only useful for
regtest and devnet. Using this on mainnet or testnet will ban
you.
.HP
\fB\-printtoconsole\fR
.IP
Send trace/debug info to console instead of debug.log file
.HP
\fB\-shrinkdebugfile\fR
.IP
Shrink debug.log file on client startup (default: 1 when no \fB\-debug\fR)
.HP
\fB\-sporkaddr=\fR<raptoreumaddress>
.IP
Override spork address. Only useful for regtest and devnet. Using this
on mainnet or testnet will ban you.
.HP
\fB\-sporkkey=\fR<privatekey>
.IP
Set the private key to be used for signing spork messages.
.HP
\fB\-uacomment=\fR<cmt>
.IP
Append comment to the user agent string
.PP
Chain selection options:
.HP
\fB\-devnet=\fR<name>
.IP
Use devnet chain with provided name
.HP
\fB\-testnet\fR
.IP
Use the test chain
.PP
Node relay options:
.HP
\fB\-bytespersigop\fR
.IP
Minimum bytes per sigop in transactions we relay and mine (default: 20)
.HP
\fB\-datacarrier\fR
.IP
Relay and mine data carrier transactions (default: 1)
.HP
\fB\-datacarriersize\fR
.IP
Maximum size of data in data carrier transactions we relay and mine
(default: 83)
.HP
\fB\-minrelaytxfee=\fR<amt>
.IP

doc/release-notes.md
Raptoreum Core version 0.15
======================

Release is now available from:

  <https://www.raptoreum.org/downloads/#wallets>

This is a new major version release, bringing new features, various bugfixes
and other improvements.

This release is mandatory for all nodes.

Please report bugs using the issue tracker at github:

  <https://github.com/raptoreum/raptoreum/issues>


Upgrading and downgrading
=========================

How to Upgrade
--------------

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over /Applications/Raptoreum-Qt (on Mac) or
raptoreumd/raptoreum-qt (on Linux). If you upgrade after DIP0003 activation and you were
using version < 0.13 you will have to reindex (start with -reindex-chainstate
or -reindex) to make sure your wallet has all the new data synced. Upgrading
from version 0.13 should not require any additional actions.

When upgrading from a version prior to 0.14.0.3, the
first startup of Raptoreum Core will run a migration process which can take a few minutes
to finish. After the migration, a downgrade to an older version is only possible with
a reindex (or reindex-chainstate).

Downgrade warning
-----------------

### Downgrade to a version < 0.14.0.3

Downgrading to a version older than 0.14.0.3 is no longer supported due to
changes in the "evodb" database format. If you need to use an older version,
you must either reindex or re-sync the whole chain.

### Downgrade of smartnodes to < 0.16

Starting with this release, masternodes will verify the protocol version of other
masternodes. This will result in PoSe punishment/banning for outdated masternodes,
so downgrading is not recommended.

Notable changes
===============

Block Reward Reallocation
-------------------------
This version implements Block Reward Reallocation which was proposed in order
to slow the growth rate of Dash’s circulating supply by encouraging the
formation of smartnodes and was voted in by the network. The resulting allocation
will split all non-proposal block rewards 40% toward miners and 60% toward
smartnodes in the end-state once the transition period is complete.

The reallocation will take place over 4.5 years with a total of 18 reallocation
periods between the start and end state. The transition is being made gradually
to avoid market volatility and minimize network disruption.

Note that this is a hardfork which must be activated by miners. To do this they
should start creating blocks signalling bit 5 in the `version` field of the block header.

### Reallocation periods

Each reallocation period will last three superblock cycles (approximately one
quarter). The following table shows the split of the non-proposal block rewards during
each period.

|Event|Miner|Smartnode|
|--|-----|-----|
|Now|50.0%|50.0%|
|1 |48.7%|51.3%|
|2 |47.4%|52.6%|
|3 |46.7%|53.3%|
|4 |46.0%|54.0%|
|5 |45.4%|54.6%|
|6 |44.8%|55.2%|
|7 |44.3%|55.7%|
|8 |43.8%|56.2%|
|9 |43.3%|56.7%|
|10|42.8%|57.2%|
|11|42.3%|57.7%|
|12|41.8%|58.2%|
|13|41.5%|58.5%|
|14|41.2%|58.8%|
|15|40.9%|59.1%|
|16|40.6%|59.4%|
|17|40.3%|59.7%|
|18|40.1%|59.9%|
|End|40.0%|60.0%|

Dynamic Activation Thresholds
-----------------------------
In Dash we have used lower thresholds (80% vs 95% in BTC) to activate upgrades
via a BIP9-like mechanism for quite some time. While it's preferable to have as much
of the network hashrate signal update readiness as possible, this can result in
quite lengthy upgrades if one large non-upgraded entity stalls
all progress. Simply lowering thresholds even further can result in network
upgrades occurring too quickly and potentially introducing network instability. This version
implements BIP9-like dynamic activation thresholds which drop from some initial
level to a minimally acceptable one over time at an increasing rate. This provides
a safe non-blocking way of activating proposals.

This mechanism applies to the Block Reward Reallocation proposal mentioned above.
Its initial threshold is 80% and it will decrease to a minimum of 60% over the
course of 10 periods. Each period is 4032 blocks (approximately one week).

Concentrated Recovery
---------------------
In the current system, signature shares are propagated to all LLMQ members
until one of them has collected enough shares to recover the signature. All
members keep propagating and verifying each share until this recovered signature
is propagated in the LLMQ. This causes significant load on the LLMQ and results
in decreased throughput.

This new system initially sends all shares to a single deterministically selected node,
so that this node can recover the signature and propagate the recovered signature.
This way only the recovered signature needs to be propagated and verified by all
members. After sending their share to this node, each member waits for a
timeout and then sends their share to another deterministically selected member.
This process is repeated until a recovered signature is finally created and propagated.

This timeout begins at two seconds and increases exponentially up to ten seconds
(i.e. `2,4,8,10,10`) for each node that times out. This is to minimize the time
taken to generate a signature if the recovery node is down, while also
minimizing the traffic generated when the network is under stress.

The new system is activated with the newly added `SPORK_21_QUORUM_ALL_CONNECTED`
which has two hardcoded values with a special meaning: `0` activates Concentrated
Recovery for every LLMQ and `1` excludes `400_60` and `400_85` quorums.

Increased number of smartnode connections
------------------------------------------
To implement "Concentrated Recovery", it is now necessary for all members of a LLMQ
to connect to all other members of the same LLMQ. This significantly increases the general
connection count for masternodes. Although these intra-quorum connections are less resource

src/raptoreumd.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2015 The Bitcoin Core developers
// Copyright (c) 2014-2021 The Dash Core developers
// Copyright (c) 2020-2023 The Raptoreum developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#if defined(HAVE_CONFIG_H)
#include <config/raptoreum-config.h>
#endif

#include <chainparams.h>
#include <clientversion.h>
#include <compat.h>
#include <fs.h>
#include <interfaces/chain.h>
#include <rpc/server.h>
#include <init.h>
#include <node/context.h>
#include <noui.h>
#include <shutdown.h>
#include <ui_interface.h>
#include <util/ref.h>
#include <util/system.h>
#include <util/strencodings.h>
#include <util/threadnames.h>
#include <stacktraces.h>

#include <functional>
#include <stdio.h>

const std::function<std::string(const char *)> G_TRANSLATION_FUN = nullptr;

static void WaitForShutdown(NodeContext &node) {
    while (!ShutdownRequested()) {
        UninterruptibleSleep(std::chrono::milliseconds{200});
    }
    Interrupt(node);
}

static bool AppInit(int argc, char *argv[]) {
    NodeContext node;

    bool fRet = false;

    util::ThreadSetInternalName("init");

    //
    // Parameters
    //
    // If Qt is used, parameters/raptoreum.conf are parsed in qt/raptoreum.cpp's main()
    SetupServerArgs();
    std::string error;
    if (!gArgs.ParseParameters(argc, argv, error)) {
        return InitError(strprintf("Error parsing command line arguments: %s\n", error));
    }

    if (gArgs.IsArgSet("-printcrashinfo")) {
        std::cout << GetCrashInfoStrFromSerializedStr(gArgs.GetArg("-printcrashinfo", "")) << std::endl;
        return true;
    }

    // Process help and version before taking care about datadir
    if (HelpRequested(gArgs) || gArgs.IsArgSet("-version")) {
        std::string
        strUsage = PACKAGE_NAME
        " Daemon version " + FormatFullVersion() + "\n";

        if (gArgs.IsArgSet("-version")) {
            strUsage += FormatParagraph(LicenseInfo()) + "\n";
        } else {
            strUsage += "\nUsage: raptoreumd [options]           Start "
            PACKAGE_NAME
            " Daemon\n";
            strUsage += "\n" + gArgs.GetHelpMessage();
        }

        tfm::format(std::cout, "%s", strUsage);
        return true;
    }

    util::Ref context{node};
    try {
        if (!CheckDataDirOption()) {
            return InitError(strprintf("Error: Specified data directory \"%s\" does not exist.\n",
                                       gArgs.GetArg("-datadir", "")));
        }

        if (!gArgs.ReadConfigFiles(error, true)) {
            return InitError(strprintf("Error reading configuration file: %s\n", error));
        }
        // Check for -testnet or -regtest parameter (Params() calls are only valid after this clause)
        try {
            SelectParams(gArgs.GetChainName());
        } catch (const std::exception &e) {
            return InitError(strprintf("%s\n", e.what()));
        }

        // Error out when loose non-argument tokens are encountered on command line
        for (int i = 1; i < argc; i++) {
            if (!IsSwitchChar(argv[i][0])) {
                return InitError(strprintf(
                        "Error: Command line contains unexpected token '%s', see raptoreumd -h for a list of options.\n",
                        argv[i]));
            }
        }

        // -server defaults to true for raptoreumd but not for the GUI so do this here
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup()) {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitParameterInteraction()) {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitSanityChecks()) {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (gArgs.GetBoolArg("-daemon", false)) {
#if HAVE_DECL_DAEMON
#if defined(MAC_OSX)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#endif
            tfm::format(std::cout, PACKAGE_NAME "daemon starting\n");

            // Daemonize
            if (daemon(1, 0)) { // don't chdir (1), do close FDs (0)
                return InitError(strprintf("daemon() failed: %s\n", strerror(errno)));
            }
#if defined(MAC_OSX)
#pragma GCC diagnostic pop
#endif
#else
            return InitError("-daemon is not supported on this operating system\n");
#endif // HAVE_DECL_DAEMON
        }
        // Lock data directory after daemonization
        if (!AppInitLockDataDirectory()) {
            // If locking the data directory failed, exit immediately
            return false;
        }
        fRet = AppInitInterfaces(node) && AppInitMain(context, node);
    } catch (...) {
        PrintExceptionContinue(std::current_exception(), "AppInit()");
    }

    if (fRet) {
        WaitForShutdown(node);
    }
    Interrupt(node);
    Shutdown(node);

    return fRet;
}

int main(int argc, char *argv[]) {
    RegisterPrettyTerminateHander();
    RegisterPrettySignalHandlers();

#ifdef WIN32
    util::WinCmdLineArgs winArgs;
    std::tie(argc, argv) = winArgs.get();
#endif
    SetupEnvironment();

    // Connect raptoreumd signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
}